validator/sawtooth_validator/journal/block_sender.py

```diff
@@ -37,6 +40,24 @@ def __init__(self, completer, gossip):
         self._completer = completer
         self._gossip = gossip
 
-    def send(self, block):
-        self._gossip.broadcast_block(block)
+    def send(self, block, injected_batch_ids=None):
+        self._gossip.broadcast_block(
+            self._remove_batches(block, keep_batches=injected_batch_ids))
         self._completer.add_block(block)
+
+    def _remove_batches(self, block, keep_batches):
+        """Returns a copy of the block without the non-injected
+        batches, which other validators are likely to already
+        have. This reduces the size of the broadcasted block.
+        """
+        block_header = BlockHeader()
+        block_header.ParseFromString(block.header)
```

Comment:
This introduces a redundant serialization/deserialization step, which will affect performance. I would probably do this clone manually rather than use the block builder since it isn't a lot of work.

```
clone = Block()
clone.header = block.header
clone.header_signature = block.header_signature
clone.batches = [batch for batch in block.batches if batch.header_signature in keep_batches]
return clone
```